### PR TITLE
Add Exchange before GroupId to improve Partial Aggregation

### PR DESCRIPTION
**Pull request by findepi**
The rule brings significant improvement in TPC-DS Q22 and Q67 while not
causing much regression in other TPC-H, TPC-DS queries. (The only
observably regressing queries were still much better than non-CBO
baseline.)

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+            double aggregationMemoryRequirements = estimateAggregationMemoryRequirements(groupingKeys, groupId, groupingSetHistogram, context);
+            if (isNaN(aggregationMemoryRequirements) || aggregationMemoryRequirements < maxPartialAggregationMemoryUsage.toBytes()) {
+                // Aggregation will be effective even without exchanges.
+                return Optional.empty();
+            }
+
+            List<Symbol> desiredHashSymbols = groupingSetHistogram.entrySet().stream()
+                    // Take only frequently used symbols
+                    .filter(entry -> entry.getCount() >= groupId.getGroupingSets().size() * GROUPING_SETS_SYMBOL_REQUIRED_FREQUENCY)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java]: what if all symbols are less frequent, shouldn't we return Optional.empty() here (short circuit)?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 2]
```diff
@@ -95,6 +95,7 @@
     public static final String LEGACY_ROW_FIELD_ORDINAL_ACCESS = "legacy_row_field_ordinal_access";
     public static final String ITERATIVE_OPTIMIZER = "iterative_optimizer_enabled";
     public static final String ITERATIVE_OPTIMIZER_TIMEOUT = "iterative_optimizer_timeout";
+    public static final String ENABLE_FORCED_EXCHANGE_BELOW_GROUP_ID = "enable_forced_exchange_below_group_id";
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Maybe `add_exchange_before_group_id`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 3]
```diff
@@ -108,6 +108,7 @@
     private double spillMaxUsedSpaceThreshold = 0.9;
     private boolean iterativeOptimizerEnabled = true;
     private boolean enableStatsCalculator = true;
+    private boolean enableForcedExchangeBelowGroupId = true;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Can we have it disabled by default?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 4]
```diff
@@ -484,6 +498,11 @@ public PlanOptimizers(
                         new PushPartialAggregationThroughJoin(),
                         new PushPartialAggregationThroughExchange(metadata.getFunctionRegistry()),
                         new PruneJoinColumns())));
+        builder.add(new IterativeOptimizer(
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: This rule uses cost model. Cost model loses estimates when there are any partial aggregation in between.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+ *           - LocalExchange
+ *             - RemoteExchange
+ * </pre>
+ * <p>
+ * Rationale: GroupId increases number of rows (number of times equal to number of grouping sets) and then
+ * partial aggregation reduces number of rows. However, under certain conditions, exchanging the rows before
+ * GroupId (before multiplication) makes partial aggregation more effective, resulting in less data being
+ * exchanged afterwards.
+ */
+public class AddExchangesBelowPartialAggregationOverGroupIdRuleSet
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java]: Maybe just `AddExchangesBeforeGroupId`?

**Pull request closed without merging**